### ISSUE TITLE
Suddenly unable to attach and debug dotnet core application in docker container

### ISSUE THREAD

**Issue by adamkarb** [microsoft/vscode-docker#2994]
I have an existing launch configuration that I have been using to successfully debug a dotnet core application within a Docker container. The configuration I use in `launch.json` is as follows:

```
{
    "name": "Docker .NET Core Attach",
    "type": "docker",
    "request": "attach",
    "platform": "netCore",
    "processId": "${command:pickRemoteProcess}",
    "sourceFileMap": {
        "/app": "${workspaceFolder}/repositories/myproject"
    }
}
```

What is happening as of late suddenly, is that when launching the debug process, the following occurs:

1. The dropdown asks me which containers I wish to attach. I select my dotnet core API container.
2. If vsdbg does not exist within the container, choose to install. Works fine.
3. Immediately afterwards, I get a vscode popup error `No process with the specified id is currently running.`

Before I began experiencing this bug/error, the `"${command:pickRemoteProcess}"` would provide me a dropdown of all processes running within the container I chose in step 1, but now it is not doing this, but immediately erroring.

What works:
If I start my dotnet core application in the container, enter the container and find the process ID within, and input it into the `processId` field of my launch configuration, it will successfully attach.

Are my configurations incorrect suddenly? Is there a better way to configure this setup?

**Comment by adamkarb** [comment]
Further context: I found the output from the `OUTPUT` pane of vscode for the `remote-attach` command. Not sure yet how to diagnose yet.

```
Executing: docker exec -i API "sh -s" < "/Users/adam/.vscode/extensions/ms-dotnettools.csharp-1.23.12/scripts/remoteProcessPickerScript"
OCI runtime exec failed: exec failed: container_linux.go:367: starting container process caused: exec: "sh -s": executable file not found in $PATH: unknown
Error Message: Command failed: docker exec -i API "sh -s" < "/Users/adam/.vscode/extensions/ms-dotnettools.csharp-1.23.12/scripts/remoteProcessPickerScript"
```

**Comment by karolz-ms** [comment]
@adamkarb could you add repro steps? E.g. does the problem reproduce with vanilla web API app created with `dotnet new webapi --no-https`?

Also, what is the OS, VS Code, and Docker extension version?

**Comment by adamkarb** [comment]
@karolz-ms System and versioning information:

Docker version 20.10.7, build f0df350

VS Code version: Code 1.56.2 (Universal) (054a9295330880ed74ceaedda236253b4f39a335, 2021-05-12T17:44:30.902Z)
OS version: Darwin x64 20.3.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz (16 x 2300)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|32.00GB (0.07GB free)|
|Process Argv|--crash-reporter-id 8641fbb2-0011-46a9-817d-eda5a1e01669|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (7)</summary>

Extension|Author (truncated)|Version
---|---|---
dotnet-test-explorer|for|0.7.5
auto-using|Fud|0.7.15
vscode-docker|ms-|1.13.0
csharp|ms-|1.23.12
mssql|ms-|1.10.1
remote-containers|ms-|0.177.2
vscodeintellicode|Vis|1.2.14


</details>

**Comment by adamkarb** [comment]
@karolz-ms I just tested with a plain web application in a docker container. I got the exact same results. The `remoteProcessPickerScript` is somehow not being picked up / executed as indicated by that log output above.

**Comment by karolz-ms** [comment]
Thanks, Adam. I can reproduce as well.

**Comment by karolz-ms** [comment]
OK I think I have figured it out. There was a relatively recent change to C# VS Code tools that changed the way process resolution works: https://github.com/OmniSharp/omnisharp-vscode/pull/4509

In most cases you won't need to specify the process picker anymore; we default to looking for `dotnet` process so it should "just work". For reference, this configuration works for me with the plain web API:

```json
{
    "name": "Docker .NET Core Attach",
    "type": "docker",
    "request": "attach",
    "platform": "netCore",
    "sourceFileMap": {
        "/src": "${workspaceFolder}"
    }
}
```

Please let us know if this gets you unblocked

**Comment by adamkarb** [comment]
@karolz-ms I am still suffering from this problem.

The error I get now looks like

```
Multiple processes were found matching the process name. Attach by process id instead.
```

There can be multiple dotnet processes running in the same container. Just in my example, for development, my entrypoint is a `dotnet run <project>` command which will take care of the compilation and execution. Because the `pickRemoteProcess` choice has been taken away from me, my only recourse without changing my development environment is to

1. Boot container.
2. Enter container and list processes to find out which process ID I need.
3. Go back to `launch.json` to hardcode it.

This is the return of `ps aux` in my container

```
root         1  0.0  0.6 3064072 80316 ?       SLsl 19:35   0:02 dotnet run ./myproject --no-launch-profile --configfile ./
root        48  0.1  0.0      0     0 ?        Z    19:35   0:04 [dotnet] <defunct>
root        65  0.0  0.0      0     0 ?        Z    19:35   0:02 [dotnet] <defunct>
root        85  0.0  0.0      0     0 ?        Z    19:35   0:01 [dotnet] <defunct>
root       160  0.0  0.0   4284   744 pts/0    Ss+  19:35   0:00 /bin/sh
root       175 10.2  0.0      0     0 ?        Z    19:35   6:45 [dotnet] <defunct>
root       395  1.6  5.7 19977160 710644 ?     SLl  19:36   1:04 dotnet exec /app/myproject/bin/container/Debug/netcoreapp2
```

Why has the ability to pick a remote process been removed completely? Can't a default be supported as well as a user choice?

**Comment by karolz-ms** [comment]
BTW opening the same issue in multiple repos like this https://github.com/microsoft/vscode/issues/125878 is not going to get you a faster resolution, or win you any friends, frankly

**Comment by karolz-ms** [comment]
I think we will need some changes on the Docker extension side to make it work with the new C# tooling. Investigating...

**Comment by adamkarb** [comment]
> BTW opening the same issue in multiple repos like this [microsoft/vscode#125878](https://github.com/microsoft/vscode/issues/125878) is not going to get you a faster resolution, or win you any friends, frankly

I opened that one first, before I realized this was the proper repository for the issue. I went ahead and closed it just for you.

**Comment by karolz-ms** [comment]
Thanks! I came to believe this is a regression in C# VS Code tooling and I have opened https://github.com/OmniSharp/omnisharp-vscode/issues/4607 for the C# tooling owners to take a look

I also noticed the default .NET image does not contain `ps` executable, so I submitted a PR to switch to Ubuntu-based images as a workaround: https://github.com/microsoft/vscode-docker/pull/2996 Without `ps` the process picker will not be able to enumerate processes in the container.

**Comment by adamkarb** [comment]
@karolz-ms 
Appreciate your diligence on this, will be following closely.

**Comment by karolz-ms** [comment]
The next OmniSharp release should have the fix for this debugging scenario--I do not think there is anything on the Docker extension side to change. Resolving as "external" for now, but please reopen/create new issue as necessary if something still does not work.
